### ISSUE TITLE
[RLlib] step() function called too early, could lead to inconsistencies

### ISSUE THREAD

**Issue by PhilippWillms** [ray-project/ray#44290]
### What happened + What you expected to happen

Hello Ray team,

deep-diving on a nested error in my custom envrionment, it seems that I have found a general topic in `ray.rllib.evaluation.env_runnver_v2.EnvRunnerV2.step()`. Using the reproduction script and adding some strategic console prints in the aforementioned EnvRunner method, I get the following relevant output. 

Here is what I understand:
1. RolloutWorker 15576 instantiates MyEnv envrionment and immediately makes a call to the env.step() function with an arbitrary, yet feasible action id.
2. Afterwards, the RolloutWorker 15576 executes a full episode with action sequence 0 -> 2 -> 1, until `terminated` flag turns true.

I see a risk that this initial call to `step()` creates inconsistency in the underlying environment instance.

> (RolloutWorker pid=15576) C:\Users\Philipp\anaconda3\envs\torch-gpu-310\lib\site-packages\gymnasium\core.py:311: 
> UserWarning: WARN: env.single_observation_space to get variables from other wrappers is deprecated and will be removed in > v1.0, to get this variable you can do `env.unwrapped.single_observation_space` for environment variables or 
> `env.get_wrapper_attr('single_observation_space')` that will search the reminding wrappers.
> (RolloutWorker pid=15576)   logger.warn(
> (RolloutWorker pid=15576) C:\Users\Philipp\anaconda3\envs\torch-gpu-310\lib\site-packages\gymnasium\core.py:311: UserWarning: WARN: env.single_action_space to get variables from other wrappers is deprecated and will be removed in v1.0, to get this variable you can do `env.unwrapped.single_action_space` for environment variables or env.get_wrapper_attr('single_action_space')` that will search the reminding wrappers.
> (RolloutWorker pid=15576)   logger.warn(
> (RolloutWorker pid=15576) Init method called.
> (RolloutWorker pid=15576) Reset method called.
> (RolloutWorker pid=15576) Reset method called.
> (RolloutWorker pid=15576) Step function called with action 2.
> (RolloutWorker pid=15576) Action key added to dict.
> (RolloutWorker pid=15576) Existing value in dict: 1
> (RolloutWorker pid=15576) State after step: [[0. 0. 0. 0.]
> (RolloutWorker pid=15576)  [0. 0. 0. 0.]
> (RolloutWorker pid=15576)  [1. 0. 0. 0.]].
> (RolloutWorker pid=25516) 2024-03-25 23:33:58,842       WARNING env.py:162 -- Your env doesn't have a .spec.max_episode_steps attribute. Your horizon will default to infinity, and your environment will not be reset.
> 2024-03-25 23:34:00,337 INFO trainable.py:164 -- Trainable.setup took 11.529 seconds. If your trainable is slow to initialize, consider setting reuse_actors=True to reduce actor creation overheads.
> (RolloutWorker pid=15576) calling step in envrunnerv2 at iteration 1
> (RolloutWorker pid=15576) terminateds BEFORE env.step() was executed {0: {'__all__': False}}
> (RolloutWorker pid=15576) active_envs {0}
> (RolloutWorker pid=15576) outputs after _process_observations []
> (RolloutWorker pid=15576) actions_to_send calculated by _process_policy_eval_results:  defaultdict(<class 'dict'>, {0: {'agent0': 0}})
> (RolloutWorker pid=15576) Having sent actions to env
> (RolloutWorker pid=15576) calling step in envrunnerv2 at iteration 2
> (RolloutWorker pid=15576) terminateds BEFORE env.step() was executed {0: {'__all__': False}}
> (RolloutWorker pid=15576) active_envs {0}
> (RolloutWorker pid=15576) outputs after _process_observations []
> (RolloutWorker pid=15576) actions_to_send calculated by _process_policy_eval_results:  defaultdict(<class 'dict'>, {0: {'agent0': 2}})
> (RolloutWorker pid=15576) Having sent actions to env
> (RolloutWorker pid=15576) calling step in envrunnerv2 at iteration 3
> (RolloutWorker pid=15576) terminateds BEFORE env.step() was executed {0: {'__all__': False}}
> (RolloutWorker pid=15576) active_envs {0}
> (RolloutWorker pid=15576) outputs after _process_observations []
> (RolloutWorker pid=15576) actions_to_send calculated by _process_policy_eval_results:  defaultdict(<class 'dict'>, {0: {'agent0': 1}})
> (RolloutWorker pid=15576) Having sent actions to env
> (RolloutWorker pid=15576) calling step in envrunnerv2 at iteration 4
> (RolloutWorker pid=15576) terminateds BEFORE env.step() was executed {0: {'__all__': True}}
> (RolloutWorker pid=15576) active_envs {0}
> (RolloutWorker pid=15576) outputs after _process_observations [RolloutMetrics(episode_length=3, episode_reward=-3.0, agent_rewards={('agent0', 'default_policy'): -3.0}, custom_metrics={}, perf_stats={}, hist_data={}, media={}, episode_faulty=False, connector_metrics={'default_policy': {'agent_connectors': {'StateBufferConnector_ms': 0.0, 'ViewRequirementAgentConnector_ms': 0.5167722702026367}, 'action_connectors': {'StateBufferConnector_ms': 0.0, 'ViewRequirementAgentConnector_ms': 0.5167722702026367}}})]

### Versions / Dependencies

ray==2.10.0
gymnasium==0.29.1
torch==2.2.1+cu121
Windows == 10.0.22631 Build 22631

### Reproduction script

from pprint import PrettyPrinter
from typing import OrderedDict, Tuple
import gymnasium
import numpy as np
from gymnasium.spaces import Box, Discrete, Dict
from gymnasium.wrappers import TransformObservation
import ray
from ray.rllib.algorithms import PPOConfig
from ray.rllib.core.rl_module.rl_module import SingleAgentRLModuleSpec
from ray.rllib.examples.rl_module.action_masking_rlm import TorchActionMaskRLM
from ray.tune.registry import register_env

class MyRealObsWrapper(TransformObservation):
    """Special Wrapper needed for new RLlib API stack."""

    def __init__(self, env):
        super().__init__(env, self.__transform)

    def __transform(self, orig_obs):
        new_obs = orig_obs
        for b in new_obs.keys():
            if b not in ["static_features"]:
                new_obs[b] = np.reshape(new_obs[b], -1)
        # Important to update the observation space, otherwise the RLlib algorithms will not work
        self.observation_space["observations"] = Box(0, 1, (len(new_obs["observations"]),))
        return new_obs

class MyEnv(gymnasium.Env):
    def __init__(self, *args, **kwargs):
        print("Init method called.")        
        self.action_space = Discrete(3)
        self.observation_space = Dict(
            {
                "action_mask": Box(
                    low=0, high=1, shape=(self.action_space.n,), dtype=np.int8
                ),
                "observations": Box(
                    low=0.0,
                    high=1.0,
                    shape=(3, 4),
                    dtype=np.float32,
                ),
                # "static_features": Dict(...)
            }
        )
        self.episode_done = False
        self._action_max_helper = np.ones(self.action_space.n, dtype=np.int8)
        self.state = np.zeros((3, 4), dtype=np.float32)
            
    def step(self, action: int) -> Tuple[OrderedDict, float, bool, bool, dict]:
        print(f"Step function called with action {action}.")
        # Error handling for invalid action
        if (action < 0) | (action > self.action_space.n):
            e_string = f"Action [{action}] is not valid! Size of the action space: [{self.action_space.n}]."
            raise ValueError(e_string)
        if self._action_max_helper[action] == 0:
            e_string = f"Action [{action}] is not valid as chosen already in episode !"
            raise ValueError(e_string)
        
        some_dict = {}
        if action not in some_dict.keys():
            some_dict[action] = 1
            print(f"Action key added to dict.")
        print(f"Existing value in dict: {some_dict[action]}")
        
        reward = 0 - action
        self.state[action][0] = 1
        self._action_max_helper[action] = 0
        if all(self._action_max_helper[k] == 0 for k in range(3)):
            self.episode_done = True
        print(f"State after step: {self.state}.")
        return self._get_state_repr(), reward, self.episode_done, False, {}
    
    def _get_state_repr(self) -> OrderedDict:
        return {
            "action_mask": self._action_max_helper,
            "observations": self.state,
        }

    def reset(self, *, seed=None, options=None) -> Tuple[OrderedDict, dict]:
        self.episode_done = False
        # Initial state representation = shape of the obs space.
        self.state = np.zeros((3, 4), dtype=np.float32)
        # Initial action mask = all actions are allowed.
        self._action_max_helper = np.ones(self.action_space.n, dtype=np.int8)
        return self._get_state_repr(), {}

def env_creator(env_config):
    env = MyEnv()
    env = MyRealObsWrapper(env)
    return env

# Use classic API to register environment
register_env("myenv_wrapped", env_creator)

if __name__ == "__main__":
    rlm_spec = SingleAgentRLModuleSpec(module_class=TorchActionMaskRLM)

    # Algorithm Config, but with the latest RLlib API
    config = (
        PPOConfig().environment("myenv_wrapped")
        # We need to disable preprocessing of observations, because preprocessing
        # would flatten the observation dict of the environment.
        .experimental(_disable_preprocessor_api=True, _enable_new_api_stack=True)
        .framework("torch")
        .resources(
            num_gpus=1, num_cpus_per_worker=2, num_gpus_per_worker=0.3
        )
        .rl_module(rl_module_spec=rlm_spec)
        .training(lr=1e-3, train_batch_size=50, sgd_minibatch_size=10)
    )

    algo = config.build()

    # run manual training loop and print results after each iteration
    for _ in range(10):
        result = algo.train()
        print(PrettyPrinter(result))

    ray.shutdown()

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by PhilippWillms** [comment]
This was a temporary issue during the early days when development on new API stack started and interfered with the old API stack.
